cmd/kops/toolbox_instance_selector.go

```diff
@@ -0,0 +1,487 @@
+	clusterAutoscaler  = "cluster-autoscaler"
+	usageClassSpot     = "spot"
+	usageClassOndemand = "on-demand"
+	igName             = "instance-group-name"
+	dryRun             = "dry-run"
+	output             = "output"
+)
+
+const (
+	nameRegex = `^[a-zA-Z0-9\-_]{1,128}$`
+)
+
+// InstanceSelectorOptions is a struct representing non-filter flags passed into instance-selector
+type InstanceSelectorOptions struct {
+	NodeCountMin       int32
+	NodeCountMax       int32
+	NodeVolumeSize     *int32
+	NodeSecurityGroups []string
+	ClusterAutoscaler  bool
+	InstanceGroupName  string
+	InstanceGroupCount int
+	Output             string
+	DryRun             bool
+}
+
+var (
+	toolboxInstanceSelectorLong = templates.LongDesc(i18n.T(`
+	Generate AWS EC2 on-demand or spot instance-groups by providing resource specs like vcpus and memory rather than instance types.`))
+
+	toolboxInstanceSelectorExample = templates.Examples(i18n.T(`
+
+	## Create a best-practices spot instance-group using a MixInstancesPolicy and Capacity-Optimized spot allocation strategy
+	## --flexible defaults to a 1:2 vcpus to memory ratio and 4 vcpus
+	kops toolbox instance-selector --usage-class spot --flexible --instance-group-name my-spot-mig
+
+	## Create a best-practices on-demand instance-group with custom vcpus and memory range filters
+	kops toolbox instance-selector --instance-group-name ondemand-ig --vcpus-min=2 --vcpus-max=4 --memory-min 2048 --memory-max 4096
+	`))
+
+	toolboxInstanceSelectorShort = i18n.T(`Generate on-demand or spot instance-group specs by providing resource specs like vcpus and memory.`)
+)
+
+// NewCmdToolboxInstanceSelector defines the cobra command for the instance-selector tool
+func NewCmdToolboxInstanceSelector(f *util.Factory, out io.Writer) *cobra.Command {
+	commandline := cli.New(
+		"instance-selector",
+		toolboxInstanceSelectorShort,
+		toolboxInstanceSelectorLong,
+		toolboxInstanceSelectorExample,
+		func(cmd *cobra.Command, args []string) {},
+	)
+	commandline.Command.Run = func(cmd *cobra.Command, args []string) {
+		ctx := context.TODO()
+
+		if err := rootCommand.ProcessArgs(args); err != nil {
+			exitWithError(err)
+		}
+		err := RunToolboxInstanceSelector(ctx, f, out, rootCommand.ClusterName(), &commandline)
+		if err != nil {
+			exitWithError(err)
+		}
+	}
+
+	cpuArchs := []string{"x86_64", "i386", "arm64"}
+	cpuArchDefault := "x86_64"
+	placementGroupStrategies := []string{"cluster", "partition", "spread"}
+	usageClasses := []string{usageClassSpot, usageClassOndemand}
+	usageClassDefault := usageClassOndemand
+	outputDefault := "yaml"
+	dryRunDefault := false
+	clusterAutoscalerDefault := false
+	nodeCountMinDefault := 2
+	nodeCountMaxDefault := 15
+	maxResultsDefault := 20
+
+	commandline.StringFlag(igName, nil, nil, "Name of the Instance-Group", func(val interface{}) error {
+		if val == nil {
+			return fmt.Errorf("error you must supply --%s", igName)
+		}
+		matched, err := regexp.MatchString(nameRegex, *val.(*string))
```

Comment:
I'm wondering if we perform this validation elsewhere and if we can share the same validation logic. Or if we don't, I'd question whether we need it here. (I don't know the answer to that question, I can try to look it up in a bit)

In general for this new command I'd lean towards making it reasonably consistent with `kops create ig` logic.

Reply:
looks like validation is not performed on instance group name in `create ig` . I've removed the validation in this PR.